--- a/net/minecraft/src/BlockStone.java
+++ b/net/minecraft/src/BlockStone.java
@@ -1,13 +1,43 @@
 package net.minecraft.src;
 
 import java.util.Random;
+import java.util.stream.IntStream;
 
 public class BlockStone extends Block {
-    public BlockStone(int id, int tex) {
+    private final Type type;
+
+    public enum Type {
+        STONE,
+        COBBLESTONE,
+        POLISHED
+    }
+
+    public BlockStone(int id, int tex, Type type) {
         super(id, tex, Material.rock);
+        this.type = type;
     }
 
     public int idDropped(int metadata, Random rand) {
-        return Block.cobblestone.blockID;
+        return type == Type.STONE ? Block.cobblestone.blockID : this.blockID;
+    }
+
+    public int metadataDropped(int metadata, Random rand) {
+        return metadata;
+    }
+
+    public int[] availableMetadata() {
+        return IntStream.range(0, 3).toArray();
+    }
+
+    public int getBlockTextureFromSideAndMetadata(int side, int metadata) {
+        if (metadata == 0) {
+            return this.blockIndexInTexture;
+        }
+        switch (type) {
+            case STONE: return textureXY(8, 5 + metadata);
+            case COBBLESTONE: return textureXY(11, 5 + metadata);
+            case POLISHED: return textureXY(10, 5 + metadata);
+            default: return this.blockIndexInTexture;
+        }
     }
 }
