--- a/net/minecraft/src/BlockJukeBox.java
+++ b/net/minecraft/src/BlockJukeBox.java
@@ -20,7 +20,7 @@
     }
 
     public void ejectRecord(World worldObj, int x, int y, int z, int i5) {
-        worldObj.playRecord((String)null, x, y, z);
+        worldObj.playRecord(null, x, y, z, null);
         worldObj.setBlockMetadataWithNotify(x, y, z, 0);
         int i6 = Item.record13.shiftedIndex + i5 - 1;
         float f7 = 0.7F;
@@ -38,7 +38,7 @@
                 this.ejectRecord(worldObj, x, y, z, metadata);
             }
 
-            super.dropBlockAsItemWithChance(worldObj, x, y, z, metadata, chance);
+            super.dropBlockAsItemWithChance(worldObj, x, y, z, 0, chance);
         }
     }
 }
