--- a/net/minecraft/src/Block.java
+++ b/net/minecraft/src/Block.java
@@ -1,6 +1,9 @@
 package net.minecraft.src;
 
+import io.github.bluestaggo.mcdiverge.*;
+
 import java.util.ArrayList;
+import java.util.Objects;
 import java.util.Random;
 
 public class Block {
@@ -20,10 +23,10 @@
     public static final int[] lightOpacity = new int[256];
     public static final boolean[] canBlockGrass = new boolean[256];
     public static final int[] lightValue = new int[256];
-    public static final Block stone = (new BlockStone(1, 1)).setHardness(1.5F).setResistance(10.0F).setStepSound(soundStoneFootstep);
+    public static final Block stone = (new BlockStone(1, 1, BlockStone.Type.STONE)).setHardness(1.5F).setResistance(10.0F).setStepSound(soundStoneFootstep);
     public static final BlockGrass grass = (BlockGrass)(new BlockGrass(2)).setHardness(0.6F).setStepSound(soundGrassFootstep);
     public static final Block dirt = (new BlockDirt(3, 2)).setHardness(0.5F).setStepSound(soundGravelFootstep);
-    public static final Block cobblestone = (new Block(4, 16, Material.rock)).setHardness(2.0F).setResistance(10.0F).setStepSound(soundStoneFootstep);
+    public static final Block cobblestone = (new BlockStone(4, 16, BlockStone.Type.COBBLESTONE)).setHardness(2.0F).setResistance(10.0F).setStepSound(soundStoneFootstep);
     public static final Block planks = (new Block(5, 4, Material.wood)).setHardness(2.0F).setResistance(5.0F).setStepSound(soundWoodFootstep);
     public static final Block sapling = (new BlockSapling(6, 15)).setHardness(0.0F).setStepSound(soundGrassFootstep);
     public static final Block bedrock = (new Block(7, 17, Material.rock)).setHardness(-1.0F).setResistance(6000000.0F).setStepSound(soundStoneFootstep);
@@ -33,49 +36,50 @@
     public static final Block lavaStill = (new BlockStationary(11, Material.lava)).setHardness(100.0F).setLightValue(1.0F).setLightOpacity(255);
     public static final Block sand = (new BlockSand(12, 18)).setHardness(0.5F).setStepSound(soundSandFootstep);
     public static final Block gravel = (new BlockGravel(13, 19)).setHardness(0.6F).setStepSound(soundGravelFootstep);
-    public static final Block oreGold = (new BlockOre(14, 32)).setHardness(3.0F).setResistance(5.0F).setStepSound(soundStoneFootstep);
-    public static final Block oreIron = (new BlockOre(15, 33)).setHardness(3.0F).setResistance(5.0F).setStepSound(soundStoneFootstep);
-    public static final Block oreCoal = (new BlockOre(16, 34)).setHardness(3.0F).setResistance(5.0F).setStepSound(soundStoneFootstep);
+    public static final Block oreGold = (new BlockOre(14, 32, textureXY(8, 8))).setHardness(3.0F).setResistance(5.0F).setStepSound(soundStoneFootstep);
+    public static final Block oreIron = (new BlockOre(15, 33, textureXY(9, 8))).setHardness(3.0F).setResistance(5.0F).setStepSound(soundStoneFootstep);
+    public static final Block oreCoal = (new BlockOre(16, 34, textureXY(10, 8), 263)).setHardness(3.0F).setResistance(5.0F).setStepSound(soundStoneFootstep);
     public static final Block wood = (new BlockLog(17)).setHardness(2.0F).setStepSound(soundWoodFootstep);
     public static final BlockLeaves leaves = (BlockLeaves)(new BlockLeaves(18, 52)).setHardness(0.2F).setLightOpacity(1).setStepSound(soundGrassFootstep);
     public static final Block sponge = (new BlockSponge(19)).setHardness(0.6F).setStepSound(soundGrassFootstep);
     public static final Block glass = (new BlockGlass(20, 49, Material.glass, false)).setHardness(0.3F).setStepSound(soundGlassFootstep);
-    public static final Block clothRed = null;
-    public static final Block clothOrange = null;
-    public static final Block clothYellow = null;
-    public static final Block clothChartreuse = null;
-    public static final Block clothGreen = null;
-    public static final Block clothSpringGreen = null;
-    public static final Block clothCyan = null;
-    public static final Block clothCapri = null;
-    public static final Block clothUltramarine = null;
-    public static final Block clothViolet = null;
-    public static final Block clothPurple = null;
-    public static final Block clothMagenta = null;
-    public static final Block clothRose = null;
-    public static final Block clothDarkGray = null;
+    public static final Block polishedStone = (new BlockStone(21, 6, BlockStone.Type.POLISHED)).setHardness(2.0F).setResistance(10.0F).setStepSound(soundStoneFootstep);
+    public static final Block woodStairDouble = (new BlockStep(22, Material.wood, true, new BlockStep.Type[] {
+            new BlockStep.Type(textureXY(4, 0))
+    })).setHardness(2.0F).setResistance(5.0F).setStepSound(soundWoodFootstep);
+    public static final Block woodStairSingle = (new BlockStep(23, Material.wood, false, ((BlockStep) woodStairDouble).slabTypes)).setHardness(2.0F).setResistance(5.0F).setStepSound(soundWoodFootstep);
     public static final Block cloth = (new Block(35, 64, Material.cloth)).setHardness(0.8F).setStepSound(soundClothFootstep);
-    public static final Block clothWhite = null;
     public static final BlockFlower plantYellow = (BlockFlower)(new BlockFlower(37, 13)).setHardness(0.0F).setStepSound(soundGrassFootstep);
     public static final BlockFlower plantRed = (BlockFlower)(new BlockFlower(38, 12)).setHardness(0.0F).setStepSound(soundGrassFootstep);
     public static final BlockFlower mushroomBrown = (BlockFlower)(new BlockMushroom(39, 29)).setHardness(0.0F).setStepSound(soundGrassFootstep).setLightValue(0.125F);
     public static final BlockFlower mushroomRed = (BlockFlower)(new BlockMushroom(40, 28)).setHardness(0.0F).setStepSound(soundGrassFootstep);
     public static final Block blockGold = (new BlockOreBlock(41, 39)).setHardness(3.0F).setResistance(10.0F).setStepSound(soundMetalFootstep);
     public static final Block blockSteel = (new BlockOreBlock(42, 38)).setHardness(5.0F).setResistance(10.0F).setStepSound(soundMetalFootstep);
-    public static final Block stairDouble = (new BlockStep(43, true)).setHardness(2.0F).setResistance(10.0F).setStepSound(soundStoneFootstep);
-    public static final Block stairSingle = (new BlockStep(44, false)).setHardness(2.0F).setResistance(10.0F).setStepSound(soundStoneFootstep);
+    public static final Block stairDouble = (new BlockStep(43, Material.rock, true, new BlockStep.Type[] {
+            new BlockStep.Type(textureXY(0, 1)),
+            new BlockStep.Type(textureXY(11, 6)),
+            new BlockStep.Type(textureXY(11, 7)),
+            new BlockStep.Type(textureXY(1, 0)),
+            new BlockStep.Type(textureXY(8, 6)),
+            new BlockStep.Type(textureXY(8, 7)),
+            new BlockStep.Type(textureXY(6, 0), true),
+            new BlockStep.Type(textureXY(10, 6), true),
+            new BlockStep.Type(textureXY(10, 7), true),
+            new BlockStep.Type(textureXY(7, 0))
+    })).setHardness(2.0F).setResistance(10.0F).setStepSound(soundStoneFootstep);
+    public static final Block stairSingle = (new BlockStep(44, Material.rock, false, ((BlockStep) stairDouble).slabTypes)).setHardness(2.0F).setResistance(10.0F).setStepSound(soundStoneFootstep);
     public static final Block brick = (new Block(45, 7, Material.rock)).setHardness(2.0F).setResistance(10.0F).setStepSound(soundStoneFootstep);
     public static final Block tnt = (new BlockTNT(46, 8)).setHardness(0.0F).setStepSound(soundGrassFootstep);
     public static final Block bookshelf = (new BlockBookshelf(47, 35)).setHardness(1.5F).setStepSound(soundWoodFootstep);
     public static final Block cobblestoneMossy = (new Block(48, 36, Material.rock)).setHardness(2.0F).setResistance(10.0F).setStepSound(soundStoneFootstep);
     public static final Block obsidian = (new BlockObsidian(49, 37)).setHardness(10.0F).setResistance(2000.0F).setStepSound(soundStoneFootstep);
     public static final Block torch = (new BlockTorch(50, 80)).setHardness(0.0F).setLightValue(0.9375F).setStepSound(soundWoodFootstep);
-    public static final BlockFire fire = (BlockFire)((BlockFire)(new BlockFire(51, 31)).setHardness(0.0F).setLightValue(1.0F).setStepSound(soundWoodFootstep));
+    public static final BlockFire fire = (BlockFire) (new BlockFire(51, 31)).setHardness(0.0F).setLightValue(1.0F).setStepSound(soundWoodFootstep);
     public static final Block mobSpawner = (new BlockMobSpawner(52, 65)).setHardness(5.0F).setStepSound(soundMetalFootstep);
     public static final Block stairCompactWood = new BlockStairs(53, planks);
     public static final Block chest = (new BlockChest(54)).setHardness(2.5F).setStepSound(soundWoodFootstep);
     public static final Block redstoneWire = (new BlockRedstoneWire(55, 84)).setHardness(0.0F).setStepSound(soundPowderFootstep);
-    public static final Block oreDiamond = (new BlockOre(56, 50)).setHardness(3.0F).setResistance(5.0F).setStepSound(soundStoneFootstep);
+    public static final Block oreDiamond = (new BlockOre(56, 50, textureXY(11, 8), 264)).setHardness(3.0F).setResistance(5.0F).setStepSound(soundStoneFootstep);
     public static final Block blockDiamond = (new BlockOreBlock(57, 40)).setHardness(5.0F).setResistance(10.0F).setStepSound(soundMetalFootstep);
     public static final Block workbench = (new BlockWorkbench(58)).setHardness(2.5F).setStepSound(soundWoodFootstep);
     public static final Block crops = (new BlockCrops(59, 88)).setHardness(0.0F).setStepSound(soundGrassFootstep);
@@ -92,8 +96,8 @@
     public static final Block pressurePlateStone = (new BlockPressurePlate(70, stone.blockIndexInTexture, EnumMobType.mobs)).setHardness(0.5F).setStepSound(soundStoneFootstep);
     public static final Block doorSteel = (new BlockDoor(71, Material.iron)).setHardness(5.0F).setStepSound(soundMetalFootstep);
     public static final Block pressurePlateWood = (new BlockPressurePlate(72, planks.blockIndexInTexture, EnumMobType.everything)).setHardness(0.5F).setStepSound(soundWoodFootstep);
-    public static final Block oreRedstone = (new BlockRedstoneOre(73, 51, false)).setHardness(3.0F).setResistance(5.0F).setStepSound(soundStoneFootstep);
-    public static final Block oreRedstoneGlowing = (new BlockRedstoneOre(74, 51, true)).setLightValue(0.625F).setHardness(3.0F).setResistance(5.0F).setStepSound(soundStoneFootstep);
+    public static final Block oreRedstone = (new BlockRedstoneOre(73, 51, textureXY(12, 8), false)).setHardness(3.0F).setResistance(5.0F).setStepSound(soundStoneFootstep);
+    public static final Block oreRedstoneGlowing = (new BlockRedstoneOre(74, 51, textureXY(12, 8), true)).setLightValue(0.625F).setHardness(3.0F).setResistance(5.0F).setStepSound(soundStoneFootstep);
     public static final Block torchRedstoneIdle = (new BlockRedstoneTorch(75, 115, false)).setHardness(0.0F).setStepSound(soundWoodFootstep);
     public static final Block torchRedstoneActive = (new BlockRedstoneTorch(76, 99, true)).setHardness(0.0F).setLightValue(0.5F).setStepSound(soundWoodFootstep);
     public static final Block button = (new BlockButton(77, stone.blockIndexInTexture)).setHardness(0.5F).setStepSound(soundStoneFootstep);
@@ -105,6 +109,14 @@
     public static final Block reed = (new BlockReed(83, 73)).setHardness(0.0F).setStepSound(soundGrassFootstep);
     public static final Block jukebox = (new BlockJukeBox(84, 74)).setHardness(2.0F).setResistance(10.0F).setStepSound(soundStoneFootstep);
     public static final Block fence = (new BlockFence(85, 4)).setHardness(2.0F).setResistance(5.0F).setStepSound(soundWoodFootstep);
+    public static final Block speleothem = (new BlockSpeleothem(86, textureXY(0, 9))).setHardness(0.5F).setStepSound(soundStoneFootstep);
+    public static final Block oreRuby = (new BlockOre(87, textureXY(12, 6), textureXY(12, 7), 350, 3)).setHardness(3.0F).setResistance(5.0F).setStepSound(soundStoneFootstep);
+    public static final Block oreSapphire = (new BlockOre(88, textureXY(13, 6), textureXY(13, 7), 351, 3)).setHardness(3.0F).setResistance(5.0F).setStepSound(soundStoneFootstep);
+    public static final Block blockRuby = (new BlockOreBlock(89, textureXY(6, 7))).setHardness(5.0F).setResistance(10.0F).setStepSound(soundMetalFootstep);
+    public static final Block blockSapphire = (new BlockOreBlock(90, textureXY(7, 7))).setHardness(5.0F).setResistance(10.0F).setStepSound(soundMetalFootstep);
+    public static final Block blasterIdle = (new BlockBlaster(91, false)).setHardness(7.0F).setStepSound(soundMetalFootstep);
+    public static final Block blasterActive = (new BlockBlaster(92, true)).setHardness(7.0F).setStepSound(soundMetalFootstep).setLightValue(0.875F);
+    public static final Block refabricator = (new BlockRefabricator(93)).setHardness(10.0F).setResistance(10.0F).setStepSound(soundMetalFootstep).setLightValue(0.5F);
     public int blockIndexInTexture;
     public final int blockID;
     protected float hardness;
@@ -120,6 +132,10 @@
     public final Material material;
     public float slipperiness;
 
+    protected static int textureXY(int x, int y) {
+        return y * 16 + x;
+    }
+
     protected Block(int id, Material material) {
         this.stepSound = soundPowderFootstep;
         this.blockParticleGravity = 1.0F;
@@ -201,8 +217,12 @@
         return blockAccess.getBrightness(x, y, z);
     }
 
+    public float getBlockRenderBrightness(IBlockAccess blockAccess, int x, int y, int z) {
+        return this.getBlockBrightness(blockAccess, x, y, z) * (1 - RenderGlobal.globalBrightness) + RenderGlobal.globalBrightness;
+    }
+
     public boolean shouldSideBeRendered(IBlockAccess blockAccess, int x, int y, int z, int side) {
-        return side == 0 && this.minY > 0.0D ? true : (side == 1 && this.maxY < 1.0D ? true : (side == 2 && this.minZ > 0.0D ? true : (side == 3 && this.maxZ < 1.0D ? true : (side == 4 && this.minX > 0.0D ? true : (side == 5 && this.maxX < 1.0D ? true : !blockAccess.isBlockNormalCube(x, y, z))))));
+        return side == 0 && this.minY > 0.0D ? true : (side == 1 && this.maxY < 1.0D || (side == 2 && this.minZ > 0.0D ? true : (side == 3 && this.maxZ < 1.0D ? true : (side == 4 && this.minX > 0.0D ? true : (side == 5 && this.maxX < 1.0D ? true : !blockAccess.isBlockNormalCube(x, y, z))))));
     }
 
     public int getBlockTexture(IBlockAccess blockAccess, int x, int y, int z, int side) {
@@ -221,7 +241,7 @@
         return AxisAlignedBB.getBoundingBoxFromPool((double)x + this.minX, (double)y + this.minY, (double)z + this.minZ, (double)x + this.maxX, (double)y + this.maxY, (double)z + this.maxZ);
     }
 
-    public void getCollidingBoundingBoxes(World worldObj, int x, int y, int z, AxisAlignedBB aabb, ArrayList collidingBoundingBoxes) {
+    public void getCollidingBoundingBoxes(World worldObj, int x, int y, int z, AxisAlignedBB aabb, ArrayList<AxisAlignedBB> collidingBoundingBoxes) {
         AxisAlignedBB axisAlignedBB7 = this.getCollisionBoundingBoxFromPool(worldObj, x, y, z);
         if(axisAlignedBB7 != null && aabb.intersectsWith(axisAlignedBB7)) {
             collidingBoundingBoxes.add(axisAlignedBB7);
@@ -275,8 +295,15 @@
         return this.blockID;
     }
 
+    public int metadataDropped(int metadata, Random rand) {
+        return 0;
+    }
+
     public float blockStrength(EntityPlayer entityPlayer) {
-        return this.hardness < 0.0F ? 0.0F : (!entityPlayer.canHarvestBlock(this) ? 1.0F / this.hardness / 100.0F : entityPlayer.getCurrentPlayerStrVsBlock(this) / this.hardness / 30.0F);
+        return entityPlayer.creative ? Float.POSITIVE_INFINITY
+        : this.hardness < 0.0F ? 0.0F
+        : (!entityPlayer.canHarvestBlock(this) ? 1.0F / this.hardness / 100.0F
+        : entityPlayer.getCurrentPlayerStrVsBlock(this) / this.hardness / 30.0F);
     }
 
     public void dropBlockAsItem(World worldObj, int x, int y, int z, int metadata) {
@@ -291,11 +318,16 @@
                 if(worldObj.rand.nextFloat() <= chance) {
                     int i9 = this.idDropped(metadata, worldObj.rand);
                     if(i9 > 0) {
+                        ItemStack itemStack = new ItemStack(i9);
+                        if(i9 < 256) {
+                            itemStack.itemMetadata = this.metadataDropped(metadata, worldObj.rand);
+                        }
+
                         float f10 = 0.7F;
                         double d11 = (double)(worldObj.rand.nextFloat() * f10) + (double)(1.0F - f10) * 0.5D;
                         double d13 = (double)(worldObj.rand.nextFloat() * f10) + (double)(1.0F - f10) * 0.5D;
                         double d15 = (double)(worldObj.rand.nextFloat() * f10) + (double)(1.0F - f10) * 0.5D;
-                        EntityItem entityItem17 = new EntityItem(worldObj, (double)x + d11, (double)y + d13, (double)z + d15, new ItemStack(i9));
+                        EntityItem entityItem17 = new EntityItem(worldObj, (double)x + d11, (double)y + d13, (double)z + d15, itemStack);
                         entityItem17.delayBeforeCanPickup = 10;
                         worldObj.spawnEntityInWorld(entityItem17);
                     }
@@ -479,6 +511,20 @@
                 Item.itemsList[i0] = new ItemBlock(i0 - 256);
             }
         }
-
+    }
+
+    public int[] availableMetadata() {
+        return new int[] { 0 };
+    }
+
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Block block = (Block) o;
+        return blockID == block.blockID;
+    }
+
+    public int hashCode() {
+        return Objects.hash(blockID);
     }
 }
