--- a/net/minecraft/src/BlockFurnace.java
+++ b/net/minecraft/src/BlockFurnace.java
@@ -3,10 +3,14 @@
 import java.util.Random;
 
 public class BlockFurnace extends BlockContainer {
-    private final boolean isActive;
+    protected final boolean isActive;
 
     protected BlockFurnace(int blockID, boolean isActive) {
-        super(blockID, Material.rock);
+        this(blockID, isActive, Material.rock);
+    }
+
+    protected BlockFurnace(int blockID, boolean isActive, Material material) {
+        super(blockID, material);
         this.isActive = isActive;
         this.blockIndexInTexture = 45;
     }
@@ -91,7 +95,7 @@
         return true;
     }
 
-    public static void updateFurnaceBlockState(boolean isActive, World worldObj, int x, int y, int z) {
+    public void updateFurnaceBlockState(boolean isActive, World worldObj, int x, int y, int z) {
         int i5 = worldObj.getBlockMetadata(x, y, z);
         TileEntity tileEntity6 = worldObj.getBlockTileEntity(x, y, z);
         if(isActive) {
