--- a/net/minecraft/src/ChunkCache.java
+++ b/net/minecraft/src/ChunkCache.java
@@ -54,7 +54,8 @@
             int i6;
             if(z4) {
                 i5 = this.getBlockId(x, y, z);
-                if(i5 == Block.stairSingle.blockID || i5 == Block.tilledField.blockID) {
+                Block b = Block.blocksList[i5];
+                if((b instanceof BlockStep && !((BlockStep) b).blockType) || i5 == Block.tilledField.blockID) {
                     i6 = this.getLightValueExt(x, y + 1, z, false);
                     int i7 = this.getLightValueExt(x + 1, y, z, false);
                     int i8 = this.getLightValueExt(x - 1, y, z, false);
