--- a/net/minecraft/src/CraftingInventoryPlayerCB.java
+++ b/net/minecraft/src/CraftingInventoryPlayerCB.java
@@ -11,22 +11,26 @@
 
     public void onCraftMatrixChanged(IInventory inventory) {
         int[] i2 = new int[9];
+        int[] metaMap = new int[9];
 
         for(int i3 = 0; i3 < 3; ++i3) {
             for(int i4 = 0; i4 < 3; ++i4) {
                 int i5 = -1;
+                int metadata = -1;
                 if(i3 < 2 && i4 < 2) {
                     ItemStack itemStack6 = this.craftMatrix.getStackInSlot(i3 + i4 * 2);
                     if(itemStack6 != null) {
                         i5 = itemStack6.itemID;
+                        metadata = itemStack6.itemMetadata;
                     }
                 }
 
                 i2[i3 + i4 * 3] = i5;
+                metaMap[i3 + i4 * 3] = metadata;
             }
         }
 
-        this.craftResult.setInventorySlotContents(0, CraftingManager.getInstance().findMatchingRecipe(i2));
+        this.craftResult.setInventorySlotContents(0, CraftingManager.getInstance().findMatchingRecipe(i2, metaMap));
     }
 
     public void onCraftGuiClosed(EntityPlayer entityPlayer) {
