--- a/net/minecraft/src/CraftingInventoryWorkbenchCB.java
+++ b/net/minecraft/src/CraftingInventoryWorkbenchCB.java
@@ -6,6 +6,7 @@
 
     public void onCraftMatrixChanged(IInventory inventory) {
         int[] i2 = new int[9];
+        int[] metaMap = new int[9];
 
         for(int i3 = 0; i3 < 3; ++i3) {
             for(int i4 = 0; i4 < 3; ++i4) {
@@ -13,13 +14,15 @@
                 ItemStack itemStack6 = this.craftMatrix.getStackInSlot(i5);
                 if(itemStack6 == null) {
                     i2[i5] = -1;
+                    metaMap[i5] = -1;
                 } else {
                     i2[i5] = itemStack6.itemID;
+                    metaMap[i5] = itemStack6.itemMetadata;
                 }
             }
         }
 
-        this.craftResult.setInventorySlotContents(0, CraftingManager.getInstance().findMatchingRecipe(i2));
+        this.craftResult.setInventorySlotContents(0, CraftingManager.getInstance().findMatchingRecipe(i2, metaMap));
     }
 
     public void onCraftGuiClosed(EntityPlayer entityPlayer) {
