--- a/net/minecraft/src/EntityPlayerSP.java
+++ b/net/minecraft/src/EntityPlayerSP.java
@@ -1,20 +1,30 @@
 package net.minecraft.src;
 
+import io.github.bluestaggo.mcdiverge.GuiCreative;
+import io.github.bluestaggo.mcdiverge.GuiRecipeGuide;
+import io.github.bluestaggo.mcdiverge.GuiRefabricator;
+import io.github.bluestaggo.mcdiverge.TileEntityRefabricator;
 import net.minecraft.client.Minecraft;
 
+import java.util.regex.Pattern;
+
 public class EntityPlayerSP extends EntityPlayer {
+    private static final Pattern player4DigitPattern = Pattern.compile("Player[0-9]{0,4}");
     public MovementInput movementInput;
     private Minecraft mc;
 
     public EntityPlayerSP(Minecraft mc, World worldObj, Session session) {
         super(worldObj);
         this.mc = mc;
-        if(session != null && session.username != null && session.username.length() > 0) {
-            this.skinUrl = "http://www.minecraft.net/skin/" + session.username + ".png";
-            System.out.println("Loading texture " + this.skinUrl);
+        if (session != null && session.username != null && session.username.length() > 0) {
+            boolean useBlueStagSkin = player4DigitPattern.matcher(session.username).find();
+            if (useBlueStagSkin) {
+                this.skinUrl = "https://raw.githubusercontent.com/BlueStaggo/BlueStaggo/main/skin-mc-diverge.png";
+            } else {
+                this.skinUrl = "http://www.minecraft.net/skin/" + session.username + ".png";
+            }
+            this.username = session.username;
         }
-
-        this.username = session.username;
     }
 
     public void updateEntityActionState() {
@@ -67,6 +77,24 @@
         this.mc.displayGuiScreen(new GuiFurnace(this.inventory, furnaceTileEntity));
     }
 
+    public void displayGUIRefabricator(TileEntityRefabricator refabricatorTileEntity) {
+        this.mc.displayGuiScreen(new GuiRefabricator(this.inventory, refabricatorTileEntity));
+    }
+
+    public void displayGUIRecipeGuide() {
+        this.mc.displayGuiScreen(new GuiRecipeGuide());
+    }
+
+    public void displayGUIRecipeGuide(ItemStack itemStack) {
+        this.mc.displayGuiScreen(new GuiRecipeGuide(itemStack));
+    }
+
+    public void displayGUICreative() {
+        if (this.creative || this.mc.options.creativeDebug) {
+            this.mc.displayGuiScreen(new GuiCreative(this.inventory, this.mc));
+        }
+    }
+
     public void attackEntity(Entity entity) {
         int i2 = this.inventory.getDamageVsEntity(entity);
         if(i2 > 0) {
@@ -113,5 +141,9 @@
 
     public boolean isSneaking() {
         return this.movementInput.sneak;
+    }
+
+    public boolean isRunning() {
+        return this.movementInput.run && !this.energyRegen;
     }
 }
