--- a/net/minecraft/src/EntityTrackerEntry.java
+++ b/net/minecraft/src/EntityTrackerEntry.java
@@ -160,7 +160,7 @@
 
             if(this.trackedEntity instanceof EntityBoat) {
                 return new Packet23VehicleSpawn(this.trackedEntity, 1);
-            } else if(this.trackedEntity instanceof IAnimals) {
+            } else if(this.trackedEntity instanceof IAnimal) {
                 return new Packet24MobSpawn((EntityLiving)this.trackedEntity);
             } else {
                 throw new IllegalArgumentException("Don\'t know how to add " + this.trackedEntity.getClass() + "!");
