--- a/net/minecraft/src/EntityPickupFX.java
+++ b/net/minecraft/src/EntityPickupFX.java
@@ -32,7 +32,7 @@
         int i27 = MathHelper.floor_double(d21);
         int i28 = MathHelper.floor_double(d23 + (double)(this.yOffset / 2.0F));
         int i29 = MathHelper.floor_double(d25);
-        float f30 = this.worldObj.getBrightness(i27, i28, i29);
+        float f30 = this.worldObj.getBrightness(i27, i28, i29) * (1 - RenderGlobal.globalBrightness) + RenderGlobal.globalBrightness;
         d21 -= interpPosX;
         d23 -= interpPosY;
         d25 -= interpPosZ;
