--- a/net/minecraft/src/EntityItem.java
+++ b/net/minecraft/src/EntityItem.java
@@ -184,7 +184,16 @@
                 entityPlayer.onItemPickup(this, i2);
                 this.setEntityDead();
             }
-
         }
+    }
+
+    public float getBrightness(float renderPartialTick) {
+        return this.item.itemID == Item.starChunk.shiftedIndex ? 1.0F : super.getBrightness(renderPartialTick);
+    }
+
+    public boolean isInRangeToRenderDist(double distance) {
+        double d3 = this.boundingBox.getAverageEdgeLength();
+        d3 *= 2048.0D * this.renderDistanceWeight;
+        return distance < d3 * d3;
     }
 }
