--- a/net/minecraft/src/GuiDeleteWorld.java
+++ b/net/minecraft/src/GuiDeleteWorld.java
@@ -11,7 +11,7 @@
     }
 
     public void initButtons() {
-        this.controlList.add(new GuiButton(6, this.width / 2 - 100, this.height / 6 + 168, "Cancel"));
+        this.controlList.add(new GuiButton(106, this.width / 2 - 100, this.height / 6 + 168, "Cancel"));
     }
 
     public void selectWorld(int i1) {
@@ -19,15 +19,21 @@
         if(string2 != null) {
             this.mc.displayGuiScreen(new GuiYesNo(this, "Are you sure you want to delete this world?", "\'" + string2 + "\' will be lost forever!", i1));
         }
-
     }
 
     public void deleteWorld(boolean z1, int i2) {
         if(z1) {
             File file3 = Minecraft.getMinecraftDir();
-            World.deleteWorld(file3, this.getSaveName(i2));
+            World.deleteWorld(file3, "World" + i2);
         }
 
         this.mc.displayGuiScreen(this.parentScreen);
+    }
+
+    protected void actionPerformed(GuiButton button) {
+        if (button.id == 106)
+            this.mc.displayGuiScreen(this.parentScreen);
+        else
+            super.actionPerformed(button);
     }
 }
