--- a/net/minecraft/src/GuiScreen.java
+++ b/net/minecraft/src/GuiScreen.java
@@ -143,19 +143,25 @@
 
     }
 
+    public static long backgroundElapsedTicks = 0;
+
     public void drawBackground(int i1) {
         GL11.glDisable(GL11.GL_LIGHTING);
         GL11.glDisable(GL11.GL_FOG);
         Tessellator tessellator2 = Tessellator.instance;
-        GL11.glBindTexture(GL11.GL_TEXTURE_2D, this.mc.renderEngine.getTexture("/dirt.png"));
+        GL11.glBindTexture(GL11.GL_TEXTURE_2D, this.mc.renderEngine.getTexture("/guibg.png"));
         GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F);
-        float f3 = 32.0F;
+        float f3 = 512.0F;
+
+        float f4 = backgroundElapsedTicks / 1000.0F;
+        float f5 = f4 / 2.0F;
+
         tessellator2.startDrawingQuads();
         tessellator2.setColorOpaque_I(4210752);
-        tessellator2.addVertexWithUV(0.0D, (double)this.height, 0.0D, 0.0D, (double)((float)this.height / f3 + (float)i1));
-        tessellator2.addVertexWithUV((double)this.width, (double)this.height, 0.0D, (double)((float)this.width / f3), (double)((float)this.height / f3 + (float)i1));
-        tessellator2.addVertexWithUV((double)this.width, 0.0D, 0.0D, (double)((float)this.width / f3), (double)(0 + i1));
-        tessellator2.addVertexWithUV(0.0D, 0.0D, 0.0D, 0.0D, (double)(0 + i1));
+        tessellator2.addVertexWithUV(0.0D,            this.height,    0.0D,    f4,                                ((float)this.height / f3 + (float)i1) + f5);
+        tessellator2.addVertexWithUV(this.width,    this.height,    0.0D,    ((float)this.width / f3) + f4,    ((float)this.height / f3 + (float)i1) + f5);
+        tessellator2.addVertexWithUV(this.width,    0.0D,            0.0D,    ((float)this.width / f3) + f4,    i1 + f5);
+        tessellator2.addVertexWithUV(0.0D,            0.0D,            0.0D,    f4,                                i1 + f5);
         tessellator2.draw();
     }
 
