--- a/net/minecraft/src/ItemSpade.java
+++ b/net/minecraft/src/ItemSpade.java
@@ -1,13 +1,25 @@
 package net.minecraft.src;
 
 public class ItemSpade extends ItemTool {
-    private static Block[] blocksEffectiveAgainst = new Block[]{Block.grass, Block.dirt, Block.sand, Block.gravel, Block.snow, Block.blockSnow, Block.blockClay};
+    private static Block[] blocksEffectiveAgainst = new Block[] {
+    Block.grass,
+    Block.dirt,
+    Block.sand,
+    Block.gravel,
+    Block.snow,
+    Block.blockSnow,
+    Block.blockClay
+    };
 
     public ItemSpade(int id, int strength) {
-        super(id, 1, strength, blocksEffectiveAgainst);
+        super(id, 1, Math.abs(strength), blocksEffectiveAgainst);
+        if (strength < 0) {
+            this.efficiencyOnProperMaterial *= 2.5F;
+            this.maxDamage *= 4;
+        }
     }
 
     public boolean canHarvestBlock(Block block) {
-        return block == Block.snow ? true : block == Block.blockSnow;
+        return block == Block.snow || block == Block.blockSnow;
     }
 }
