--- a/net/minecraft/src/ItemReed.java
+++ b/net/minecraft/src/ItemReed.java
@@ -3,9 +3,9 @@
 public class ItemReed extends Item {
     private int spawnID;
 
-    public ItemReed(int id, Block block) {
+    public ItemReed(int id, int block) {
         super(id);
-        this.spawnID = block.blockID;
+        this.spawnID = block;
     }
 
     public boolean onItemUse(ItemStack itemStack, EntityPlayer entityPlayer, World worldObj, int x, int y, int z, int side) {
