--- a/net/minecraft/src/ItemAxe.java
+++ b/net/minecraft/src/ItemAxe.java
@@ -1,9 +1,35 @@
 package net.minecraft.src;
 
+import io.github.bluestaggo.mcdiverge.ItemTooltip;
+
+import java.util.stream.Stream;
+
 public class ItemAxe extends ItemTool {
-    private static Block[] blocksEffectiveAgainst = new Block[]{Block.planks, Block.bookshelf, Block.wood, Block.chest};
+    private static final Block[] blocksEffectiveAgainst = new Block[] {
+    Block.planks,
+    Block.wood,
+    Block.leaves,
+    Block.chest,
+    Block.bookshelf,
+    Block.stairCompactWood,
+    Block.workbench,
+    Block.doorWood,
+    Block.jukebox,
+    Block.fence
+    };
 
     public ItemAxe(int id, int i2) {
-        super(id, 3, i2, blocksEffectiveAgainst);
+        super(id, 3, Math.abs(i2), blocksEffectiveAgainst);
+        if (i2 < 0) {
+            this.efficiencyOnProperMaterial *= 2.5F;
+            this.maxDamage *= 4;
+        }
+    }
+
+    public Object[] getDynamicTooltip() {
+        return Stream.of(new Object[] {
+            "Break a log while crouching to fell trees.",
+            ItemTooltip.DESCRIPTION_COL
+        }, super.getDynamicTooltip()).flatMap(Stream::of).toArray();
     }
 }
