--- a/net/minecraft/src/ItemRenderer.java
+++ b/net/minecraft/src/ItemRenderer.java
@@ -20,7 +20,7 @@
         GL11.glPushMatrix();
         if(itemStack.itemID < 256 && RenderBlocks.renderItemIn3d(Block.blocksList[itemStack.itemID].getRenderType())) {
             GL11.glBindTexture(GL11.GL_TEXTURE_2D, this.mc.renderEngine.getTexture("/terrain.png"));
-            this.renderBlocksInstance.renderBlockOnInventory(Block.blocksList[itemStack.itemID]);
+            this.renderBlocksInstance.renderBlockOnInventory(Block.blocksList[itemStack.itemID], Math.max(itemStack.itemMetadata, 0));
         } else {
             if(itemStack.itemID < 256) {
                 GL11.glBindTexture(GL11.GL_TEXTURE_2D, this.mc.renderEngine.getTexture("/terrain.png"));
@@ -132,7 +132,8 @@
         GL11.glRotatef(entityPlayerSP3.prevRotationYaw + (entityPlayerSP3.rotationYaw - entityPlayerSP3.prevRotationYaw) * renderPartialTick, 0.0F, 1.0F, 0.0F);
         RenderHelper.enableStandardItemLighting();
         GL11.glPopMatrix();
-        float f4 = this.mc.theWorld.getBrightness(MathHelper.floor_double(entityPlayerSP3.posX), MathHelper.floor_double(entityPlayerSP3.posY), MathHelper.floor_double(entityPlayerSP3.posZ));
+        float f4 = this.mc.theWorld.getBrightness(MathHelper.floor_double(entityPlayerSP3.posX), MathHelper.floor_double(entityPlayerSP3.posY), MathHelper.floor_double(entityPlayerSP3.posZ))
+                   * (1 - RenderGlobal.globalBrightness) + RenderGlobal.globalBrightness;
         GL11.glColor4f(f4, f4, f4, 1.0F);
         float f5;
         float f6;
@@ -250,7 +251,7 @@
 
     private void renderWarpedTextureOverlay(float renderPartialTick) {
         Tessellator tessellator2 = Tessellator.instance;
-        float f3 = this.mc.thePlayer.getBrightness(renderPartialTick);
+        float f3 = this.mc.thePlayer.getBrightness(renderPartialTick) * (1 - RenderGlobal.globalBrightness) + RenderGlobal.globalBrightness;
         GL11.glColor4f(f3, f3, f3, 0.5F);
         GL11.glEnable(GL11.GL_BLEND);
         GL11.glBlendFunc(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA);
