--- a/net/minecraft/src/ItemBow.java
+++ b/net/minecraft/src/ItemBow.java
@@ -6,12 +6,26 @@
         this.maxStackSize = 1;
     }
 
-    public ItemStack onItemRightClick(ItemStack itemStack, World worldObj, EntityPlayer entityPlayer) {
-        if(entityPlayer.inventory.consumeInventoryItem(Item.arrow.shiftedIndex)) {
-            worldObj.playSoundAtEntity(entityPlayer, "random.bow", 1.0F, 1.0F / (rand.nextFloat() * 0.4F + 0.8F));
-            worldObj.spawnEntityInWorld(new EntityArrow(worldObj, entityPlayer));
-        }
-
-        return itemStack;
+    private boolean canShootArrow(EntityPlayer entityPlayer1) {
+        ItemStack bodyArmor = entityPlayer1.inventory.armorItemInSlot(0);
+        if (bodyArmor != null && bodyArmor.itemID == Item.quiver.shiftedIndex
+            && bodyArmor.itemDmg < bodyArmor.getMaxDamage()) {
+            bodyArmor.itemDmg += 1;
+            return true;
+        }
+
+        if (entityPlayer1.inventory.consumeInventoryItem(Item.arrow.shiftedIndex))
+            return true;
+
+        return false;
+    }
+
+    public ItemStack onItemRightClick(ItemStack itemStack1, World world2, EntityPlayer entityPlayer3) {
+        if (canShootArrow(entityPlayer3)) {
+            world2.playSoundAtEntity(entityPlayer3, "random.bow", 1.0F, 1.0F / (rand.nextFloat() * 0.4F + 0.8F));
+            world2.spawnEntityInWorld(new EntityArrow(world2, entityPlayer3));
+        }
+
+        return itemStack1;
     }
 }
