--- a/net/minecraft/src/ItemBlock.java
+++ b/net/minecraft/src/ItemBlock.java
@@ -43,7 +43,7 @@
         } else {
             if(worldObj.canBlockBePlacedAt(this.blockID, x, y, z, false)) {
                 Block block8 = Block.blocksList[this.blockID];
-                if(worldObj.setBlockWithNotify(x, y, z, this.blockID)) {
+                if(worldObj.setBlockAndMetadataWithNotify(x, y, z, this.blockID, itemStack.itemMetadata)) {
                     Block.blocksList[this.blockID].onBlockPlaced(worldObj, x, y, z, side);
                     worldObj.playSoundEffect((double)((float)x + 0.5F), (double)((float)y + 0.5F), (double)((float)z + 0.5F), block8.stepSound.getStepSound(), (block8.stepSound.getVolume() + 1.0F) / 2.0F, block8.stepSound.getPitch() * 0.8F);
                     --itemStack.stackSize;
