--- a/net/minecraft/src/MapGenCaves.java
+++ b/net/minecraft/src/MapGenCaves.java
@@ -131,14 +131,14 @@
                                         }
 
                                         if(b51 == Block.stone.blockID || b51 == Block.dirt.blockID || b51 == Block.grass.blockID) {
-                                            if(i48 < 10) {
-                                                b3[i46] = (byte)Block.lavaMoving.blockID;
-                                            } else {
+//                                            if(i48 < 10) {
+//                                                b3[i46] = (byte)Block.lavaStill.blockID;
+//                                            } else {
                                                 b3[i46] = 0;
                                                 if(z47 && b3[i46 - 1] == Block.dirt.blockID) {
                                                     b3[i46 - 1] = (byte)Block.grass.blockID;
                                                 }
-                                            }
+//                                            }
                                         }
                                     }
 
@@ -174,9 +174,9 @@
             }
 
             for(int i16 = 0; i16 < i15; ++i16) {
-                float f17 = this.rand.nextFloat() * (float)Math.PI * 2.0F;
-                float f18 = (this.rand.nextFloat() - 0.5F) * 2.0F / 8.0F;
-                float f19 = this.rand.nextFloat() * 2.0F + this.rand.nextFloat();
+                float f17 = this.rand.nextFloat() * (float) Math.PI * 2.0F;
+                float f18 = (this.rand.nextFloat() - 0.5F) / 4.0F;
+                float f19 = this.rand.nextFloat() * 6.0F;
                 this.generateCaveNode(i4, i5, b6, d9, d11, d13, f19, f17, f18, 0, 0, 1.0D);
             }
         }
