--- a/net/minecraft/src/NetServerHandler.java
+++ b/net/minecraft/src/NetServerHandler.java
@@ -195,7 +195,7 @@
         }
 
         if(i8 > 16 || z2) {
-            ItemStack itemStack9 = packet.id >= 0 ? new ItemStack(packet.id) : null;
+            ItemStack itemStack9 = packet.id >= 0 ? new ItemStack(packet.id).withMetadata(packet.metadata) : null;
             this.playerEntity.theItemInWorldManager.activeBlockOrUseItem(this.playerEntity, this.mcServer.worldMngr, itemStack9, i3, i4, i5, i6);
         }
 
@@ -210,7 +210,7 @@
     }
 
     public void registerPacket(Packet packet) {
-        logger.warning(this.getClass() + " wasn\'t prepared to deal with a " + packet.getClass());
+        logger.warning(this.getClass() + " wasn't prepared to deal with a " + packet.getClass());
         this.kickPlayer("Protocol error, unexpected packet");
     }
 
@@ -235,7 +235,7 @@
         double d2 = (double)packet.xPosition / 32.0D;
         double d4 = (double)packet.yPosition / 32.0D;
         double d6 = (double)packet.zPosition / 32.0D;
-        EntityItem entityItem8 = new EntityItem(this.mcServer.worldMngr, d2, d4, d6, new ItemStack(packet.itemID, packet.count));
+        EntityItem entityItem8 = new EntityItem(this.mcServer.worldMngr, d2, d4, d6, new ItemStack(packet.itemID, packet.count).withMetadata(packet.itemMetadata));
         entityItem8.motionX = (double)packet.rotation / 128.0D;
         entityItem8.motionY = (double)packet.pitch / 128.0D;
         entityItem8.motionZ = (double)packet.roll / 128.0D;
