--- a/net/minecraft/src/Packet5PlayerInventory.java
+++ b/net/minecraft/src/Packet5PlayerInventory.java
@@ -31,7 +31,8 @@
             if(s4 >= 0) {
                 byte b5 = dataInputStream.readByte();
                 short s6 = dataInputStream.readShort();
-                this.inventory[i3] = new ItemStack(s4, b5, s6);
+				byte b7 = dataInputStream.readByte();
+                this.inventory[i3] = new ItemStack(s4, b5, s6).withMetadata(b7);
             }
         }
 
@@ -48,6 +49,7 @@
                 dataOutputStream.writeShort((short)this.inventory[i2].itemID);
                 dataOutputStream.writeByte((byte)this.inventory[i2].stackSize);
                 dataOutputStream.writeShort((short)this.inventory[i2].itemDmg);
+				dataOutputStream.writeByte((short)this.inventory[i2].itemMetadata);
             }
         }
 
