--- a/net/minecraft/src/PlayerControllerMP.java
+++ b/net/minecraft/src/PlayerControllerMP.java
@@ -147,7 +147,7 @@
 
     public boolean onPlayerRightClick(EntityPlayer entityPlayer, World world, ItemStack itemStack, int x, int y, int z, int side) {
         this.syncCurrentPlayItem();
-        this.netClientHandler.addToSendQueue(new Packet15Place(itemStack != null ? itemStack.itemID : -1, x, y, z, side));
+        this.netClientHandler.addToSendQueue(new Packet15Place(itemStack != null ? itemStack.itemID : -1, itemStack != null ? itemStack.itemMetadata : 0, x, y, z, side));
         return super.onPlayerRightClick(entityPlayer, world, itemStack, x, y, z, side);
     }
 
