--- a/net/minecraft/src/RecipesCrafting.java
+++ b/net/minecraft/src/RecipesCrafting.java
@@ -2,8 +2,10 @@
 
 public class RecipesCrafting {
     public void addRecipes(CraftingManager craftingManager) {
-        craftingManager.addRecipe(new ItemStack(Block.chest), new Object[]{"###", "# #", "###", '#', Block.planks});
-        craftingManager.addRecipe(new ItemStack(Block.stoneOvenIdle), new Object[]{"###", "# #", "###", '#', Block.cobblestone});
-        craftingManager.addRecipe(new ItemStack(Block.workbench), new Object[]{"##", "##", '#', Block.planks});
+        craftingManager.addRecipe(new ItemStack(Block.chest), "###", "# #", "###", '#', Block.planks);
+        craftingManager.addRecipe(new ItemStack(Block.stoneOvenIdle), "###", "# #", "###", '#', Block.cobblestone);
+        craftingManager.addRecipe(new ItemStack(Block.workbench), "##", "##", '#', Block.planks);
+        craftingManager.addRecipe(new ItemStack(Block.blasterIdle), "###", "#X#", "###", '#', Item.ingotIron, 'X', Block.blockRuby);
+        craftingManager.addRecipe(new ItemStack(Block.refabricator), "xxx", "#S#", "#D#", 'x', Item.starDust, '#', Item.ingotGold, 'S', Block.blockSapphire, 'D', Block.blockDiamond);
     }
 }
