--- a/net/minecraft/src/RecipesTools.java
+++ b/net/minecraft/src/RecipesTools.java
@@ -2,7 +2,12 @@
 
 public class RecipesTools {
     private String[][] recipePatterns = new String[][]{{"XXX", " # ", " # "}, {"X", "#", "#"}, {"XX", "X#", " #"}, {"XX", " #", " #"}};
-    private Object[][] recipeItems = new Object[][]{{Block.planks, Block.cobblestone, Item.ingotIron, Item.diamond, Item.ingotGold}, {Item.pickaxeWood, Item.pickaxeStone, Item.pickaxeSteel, Item.pickaxeDiamond, Item.pickaxeGold}, {Item.shovelWood, Item.shovelStone, Item.shovel, Item.shovelDiamond, Item.shovelGold}, {Item.axeWood, Item.axeStone, Item.axeSteel, Item.axeDiamond, Item.axeGold}, {Item.hoeWood, Item.hoeStone, Item.hoeSteel, Item.hoeDiamond, Item.hoeGold}};
+    private Object[][] recipeItems = new Object[][]{
+    {Block.planks, Block.cobblestone, Item.ingotIron, Item.diamond, Item.ingotGold},
+    {Item.pickaxeWood, Item.pickaxeStone, Item.pickaxeSteel, Item.pickaxeDiamond, Item.pickaxeGold},
+    {Item.shovelWood, Item.shovelStone, Item.shovel, Item.shovelDiamond, Item.shovelGold},
+    {Item.axeWood, Item.axeStone, Item.axeSteel, Item.axeDiamond, Item.axeGold},
+    {Item.hoeWood, Item.hoeStone, Item.hoeSteel, Item.hoeDiamond, Item.hoeGold}};
 
     public void addRecipes(CraftingManager craftingManager) {
         for(int i2 = 0; i2 < this.recipeItems[0].length; ++i2) {
@@ -10,7 +15,7 @@
 
             for(int i4 = 0; i4 < this.recipeItems.length - 1; ++i4) {
                 Item item5 = (Item)this.recipeItems[i4 + 1][i2];
-                craftingManager.addRecipe(new ItemStack(item5), new Object[]{this.recipePatterns[i4], '#', Item.stick, 'X', object3});
+                craftingManager.addRecipe(new ItemStack(item5), this.recipePatterns[i4], '#', Item.stick, 'X', object3);
             }
         }
 
