--- a/net/minecraft/src/RenderItem.java
+++ b/net/minecraft/src/RenderItem.java
@@ -42,9 +42,9 @@
             GL11.glRotatef(f12, 0.0F, 1.0F, 0.0F);
             this.loadTexture("/terrain.png");
             float f27 = 0.25F;
-            if(!Block.blocksList[itemStack10.itemID].renderAsNormalBlock() && itemStack10.itemID != Block.stairSingle.blockID) {
-                f27 = 0.5F;
-            }
+//            if(!Block.blocksList[itemStack10.itemID].renderAsNormalBlock() && itemStack10.itemID != Block.stairSingle.blockID) {
+//                f27 = 0.5F;
+//            }
 
             GL11.glScalef(f27, f27, f27);
 
@@ -57,7 +57,7 @@
                     GL11.glTranslatef(f16, f17, f18);
                 }
 
-                this.itemRenderBlocks.renderBlockOnInventory(Block.blocksList[itemStack10.itemID]);
+                this.itemRenderBlocks.renderBlockOnInventory(Block.blocksList[itemStack10.itemID], Math.max(itemStack10.itemMetadata, 0));
                 GL11.glPopMatrix();
             }
         } else {
@@ -117,7 +117,7 @@
                 GL11.glRotatef(45.0F, 0.0F, 1.0F, 0.0F);
                 GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F);
                 GL11.glScalef(1.0F, 1.0F, 1.0F);
-                this.itemRenderBlocks.renderBlockOnInventory(block7);
+                this.itemRenderBlocks.renderBlockOnInventory(block7, Math.max(itemStack3.itemMetadata, 0));
                 GL11.glPopMatrix();
             } else if(itemStack3.getIconIndex() >= 0) {
                 GL11.glDisable(GL11.GL_LIGHTING);
