--- a/net/minecraft/src/RenderGlobal.java
+++ b/net/minecraft/src/RenderGlobal.java
@@ -59,6 +59,7 @@
     double prevSortZ = -9999.0D;
     public float damagePartialTime;
     int frustumCheckOffset = 0;
+    public static float globalBrightness = 0.0F;
 
     public RenderGlobal(Minecraft minecraft1, RenderEngine renderEngine2) {
         this.mc = minecraft1;
@@ -1033,15 +1034,19 @@
         ++this.frustumCheckOffset;
     }
 
-    public void playRecord(String record, int x, int y, int z) {
-        if(record != null) {
-            this.mc.ingameGUI.setRecordPlayingMessage("C418 - " + record);
+    public void playRecord(String name, int x, int y, int z, String title) {
+        if(title != null) {
+            this.mc.ingameGUI.setRecordPlayingMessage(title);
         }
 
-        this.mc.sndManager.playStreaming(record, (float)x, (float)y, (float)z, 1.0F, 1.0F);
+        this.mc.sndManager.playStreaming(name, (float)x, (float)y, (float)z, 1.0F, 1.0F);
     }
 
     public void playSound(String sound, double posX, double posY, double posZ, float volume, float pitch) {
+        if (sound.equals("ambient.cave.cave") && !this.mc.options.caveSounds) {
+			return;
+        }
+
         float f10 = 16.0F;
         if(volume > 1.0F) {
             f10 *= volume;
@@ -1111,5 +1116,17 @@
     }
 
     public void doNothingWithTileEntity(int x, int y, int z, TileEntity tileEntity) {
+    }
+
+    private boolean fullBright = false;
+
+    public void toggleFullbright() {
+        fullBright = !fullBright;
+        if (fullBright) {
+            RenderGlobal.globalBrightness = 1.0F;
+        } else {
+            RenderGlobal.globalBrightness = this.mc.options.brightness * 0.2F;
+        }
+        this.loadRenderers();
     }
 }
