--- a/net/minecraft/src/RenderSnowball.java
+++ b/net/minecraft/src/RenderSnowball.java
@@ -9,7 +9,7 @@
         GL11.glTranslatef((float)d2, (float)d4, (float)d6);
         GL11.glEnable(GL12.GL_RESCALE_NORMAL);
         GL11.glScalef(0.5F, 0.5F, 0.5F);
-        int i10 = Item.snowball.getIconIndex((ItemStack)null);
+        int i10 = entitySnowball1.iconIndex;
         this.loadTexture("/gui/items.png");
         Tessellator tessellator11 = Tessellator.instance;
         float f12 = (float)(i10 % 16 * 16 + 0) / 256.0F;
