--- a/net/minecraft/src/RenderMinecart.java
+++ b/net/minecraft/src/RenderMinecart.java
@@ -60,9 +60,9 @@
             GL11.glTranslatef(0.0F, 0.3125F, 0.0F);
             GL11.glRotatef(90.0F, 0.0F, 1.0F, 0.0F);
             if(entityMinecart1.minecartType == 1) {
-                (new RenderBlocks()).renderBlockOnInventory(Block.chest);
+                (new RenderBlocks()).renderBlockOnInventory(Block.chest, 0);
             } else if(entityMinecart1.minecartType == 2) {
-                (new RenderBlocks()).renderBlockOnInventory(Block.stoneOvenIdle);
+                (new RenderBlocks()).renderBlockOnInventory(Block.stoneOvenIdle, 0);
             }
 
             GL11.glRotatef(-90.0F, 0.0F, 1.0F, 0.0F);
