--- a/net/minecraft/src/RenderPlayer.java
+++ b/net/minecraft/src/RenderPlayer.java
@@ -1,11 +1,14 @@
 package net.minecraft.src;
 
+import io.github.bluestaggo.mcdiverge.ItemCharm;
 import org.lwjgl.opengl.GL11;
 
 public class RenderPlayer extends RenderLiving {
     private ModelBiped modelBipedMain = (ModelBiped)this.mainModel;
     private ModelBiped modelArmorChestplate = new ModelBiped(1.0F);
     private ModelBiped modelArmor = new ModelBiped(0.5F);
+    private ModelBiped modelCharmUpper = new ModelBiped(1.1F);
+    private ModelBiped modelCharmLower = new ModelBiped(0.6F);
     private static final String[] armorFilenamePrefix = new String[]{"cloth", "chain", "iron", "diamond", "gold"};
 
     public RenderPlayer() {
@@ -13,7 +16,11 @@
     }
 
     protected boolean setArmorModel(EntityPlayer entityPlayer1, int i2) {
-        ItemStack itemStack3 = entityPlayer1.inventory.armorItemInSlot(3 - i2);
+        boolean charmUpper = i2 == 5;
+        if (charmUpper) {
+            i2 = 4;
+        }
+        ItemStack itemStack3 = entityPlayer1.inventory.armorItemInSlot(4 - i2);
         if(itemStack3 != null) {
             Item item4 = itemStack3.getItem();
             if(item4 instanceof ItemArmor) {
@@ -29,6 +36,18 @@
                 modelBiped6.bipedLeftLeg.showModel = i2 == 2 || i2 == 3;
                 this.setRenderPassModel(modelBiped6);
                 return true;
+            } else if(item4 instanceof ItemCharm && i2 == 4 && ((ItemCharm) item4).textureName != null) {
+                this.loadTexture("/charm/" + ((ItemCharm) item4).textureName + (charmUpper ? "_upper.png" : "_lower.png"));
+                ModelBiped modelCharm = charmUpper ? modelCharmUpper : modelCharmLower;
+                modelCharm.bipedHead.showModel = true;
+                modelCharm.bipedHeadwear.showModel = true;
+                modelCharm.bipedBody.showModel = true;
+                modelCharm.bipedRightArm.showModel = true;
+                modelCharm.bipedLeftArm.showModel = true;
+                modelCharm.bipedRightLeg.showModel = true;
+                modelCharm.bipedLeftLeg.showModel = true;
+                this.setRenderPassModel(modelCharm);
+                return true;
             }
         }
 
@@ -144,5 +163,9 @@
 
     public void doRender(Entity entity1, double d2, double d4, double d6, float f8, float f9) {
         this.renderPlayer((EntityPlayer)entity1, d2, d4, d6, f8, f9);
+    }
+
+    protected float maxRenderPasses() {
+        return 6;
     }
 }
