--- a/net/minecraft/src/RenderTNTPrimed.java
+++ b/net/minecraft/src/RenderTNTPrimed.java
@@ -31,14 +31,14 @@
 
         f10 = (1.0F - ((float)entityTNTPrimed1.fuse - f9 + 1.0F) / 100.0F) * 0.8F;
         this.loadTexture("/terrain.png");
-        this.blockRenderer.renderBlockOnInventory(Block.tnt);
+        this.blockRenderer.renderBlockOnInventory(Block.tnt, 0);
         if(entityTNTPrimed1.fuse / 5 % 2 == 0) {
             GL11.glDisable(GL11.GL_TEXTURE_2D);
             GL11.glDisable(GL11.GL_LIGHTING);
             GL11.glEnable(GL11.GL_BLEND);
             GL11.glBlendFunc(GL11.GL_SRC_ALPHA, GL11.GL_DST_ALPHA);
             GL11.glColor4f(1.0F, 1.0F, 1.0F, f10);
-            this.blockRenderer.renderBlockOnInventory(Block.tnt);
+            this.blockRenderer.renderBlockOnInventory(Block.tnt, 0);
             GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F);
             GL11.glDisable(GL11.GL_BLEND);
             GL11.glEnable(GL11.GL_LIGHTING);
