--- a/net/minecraft/src/Session.java
+++ b/net/minecraft/src/Session.java
@@ -13,36 +13,4 @@
         this.username = string1;
         this.sessionId = string2;
     }
-
-    static {
-        registeredBlocksList.add(Block.stone);
-        registeredBlocksList.add(Block.cobblestone);
-        registeredBlocksList.add(Block.brick);
-        registeredBlocksList.add(Block.dirt);
-        registeredBlocksList.add(Block.planks);
-        registeredBlocksList.add(Block.wood);
-        registeredBlocksList.add(Block.leaves);
-        registeredBlocksList.add(Block.torch);
-        registeredBlocksList.add(Block.stairSingle);
-        registeredBlocksList.add(Block.glass);
-        registeredBlocksList.add(Block.cobblestoneMossy);
-        registeredBlocksList.add(Block.sapling);
-        registeredBlocksList.add(Block.plantYellow);
-        registeredBlocksList.add(Block.plantRed);
-        registeredBlocksList.add(Block.mushroomBrown);
-        registeredBlocksList.add(Block.mushroomRed);
-        registeredBlocksList.add(Block.sand);
-        registeredBlocksList.add(Block.gravel);
-        registeredBlocksList.add(Block.sponge);
-        registeredBlocksList.add(Block.cloth);
-        registeredBlocksList.add(Block.oreCoal);
-        registeredBlocksList.add(Block.oreIron);
-        registeredBlocksList.add(Block.oreGold);
-        registeredBlocksList.add(Block.blockSteel);
-        registeredBlocksList.add(Block.blockGold);
-        registeredBlocksList.add(Block.bookshelf);
-        registeredBlocksList.add(Block.tnt);
-        registeredBlocksList.add(Block.obsidian);
-        System.out.println(registeredBlocksList.size());
-    }
 }
