--- /dev/null
+++ b/io/github/bluestaggo/mcdiverge/BlockRefabricator.java
@@ -1,0 +1,24 @@
+package io.github.bluestaggo.mcdiverge;
+
+import net.minecraft.src.*;
+
+public class BlockRefabricator extends BlockContainer {
+    public BlockRefabricator(int id) {
+        super(id, Material.iron);
+        this.blockIndexInTexture = textureXY(8, 1);
+    }
+
+    public int getBlockTextureFromSide(int side) {
+        return side == 1 ? textureXY(15, 6) : side == 0 ? textureXY(8, 3) : textureXY(14, 6);
+    }
+
+    public boolean blockActivated(World worldObj, int x, int y, int z, EntityPlayer entityPlayer) {
+        TileEntityRefabricator tileEntityRefabricator = (TileEntityRefabricator)worldObj.getBlockTileEntity(x, y, z);
+        entityPlayer.displayGUIRefabricator(tileEntityRefabricator);
+        return true;
+    }
+
+    protected TileEntity getBlockEntity() {
+        return new TileEntityRefabricator();
+    }
+}
