--- /dev/null
+++ b/io/github/bluestaggo/mcdiverge/CreateWorldInfo.java
@@ -1,0 +1,28 @@
+package io.github.bluestaggo.mcdiverge;
+
+public class CreateWorldInfo {
+    public String name;
+    public long seed;
+    public boolean snowy;
+    public boolean creative;
+
+    public CreateWorldInfo withName(String name) {
+        this.name = name;
+        return this;
+    }
+
+    public CreateWorldInfo withSeed(long seed) {
+        this.seed = seed;
+        return this;
+    }
+
+    public CreateWorldInfo withSnow(boolean snowy) {
+        this.snowy = snowy;
+        return this;
+    }
+
+    public CreateWorldInfo withCreative(boolean creative) {
+        this.creative = creative;
+        return this;
+    }
+}
