--- /dev/null
+++ b/io/github/bluestaggo/mcdiverge/FurnaceRecipe.java
@@ -1,0 +1,26 @@
+package io.github.bluestaggo.mcdiverge;
+
+import net.minecraft.src.ItemStack;
+
+public class FurnaceRecipe {
+    private final ItemStack input, output;
+
+    public FurnaceRecipe(ItemStack input, ItemStack output) {
+        this.input = input;
+        this.output = output;
+    }
+
+    public ItemStack getInput() {
+        return input;
+    }
+
+    public ItemStack getOutput(ItemStack input) {
+        return output.itemMetadata == -1 ? output.copy().withMetadata(input.itemMetadata) : output;
+    }
+
+    public boolean canSmelt(ItemStack input) {
+        return this.input.itemID == input.itemID
+                && (this.input.itemMetadata < 0
+                    || this.input.itemMetadata == input.itemMetadata);
+    }
+}
