--- /dev/null
+++ b/io/github/bluestaggo/mcdiverge/GuiCredits.java
@@ -1,0 +1,48 @@
+package io.github.bluestaggo.mcdiverge;
+
+import net.minecraft.src.*;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.List;
+
+public class GuiCredits extends GuiScreen {
+    private final GuiScreen parent;
+    private List<String> lines = new ArrayList<>();
+
+    public GuiCredits(GuiScreen parent) {
+        this.parent = parent;
+
+        try {
+            BufferedReader bufferedReader2 = new BufferedReader(new InputStreamReader(GuiCredits.class.getResourceAsStream("/credits.txt")));
+            String string3;
+
+            while((string3 = bufferedReader2.readLine()) != null) {
+                string3 = string3.trim();
+                if(string3.length() > 0) {
+                    this.lines.add(string3);
+                }
+            }
+        } catch (IOException e) {
+            this.lines.add("Failed to get credits");
+        }
+    }
+
+    public void initGui() {
+        this.controlList.add(new GuiButton(1, this.width / 2 - 50, this.height - 30, 100, 20, "Exit"));
+    }
+
+    protected void actionPerformed(GuiButton button) {
+        this.mc.displayGuiScreen(this.parent);
+    }
+
+    public void drawScreen(int mouseX, int mouseY, float renderPartialTick) {
+        this.drawDefaultBackground();
+        for (int i = 0; i < lines.size(); i++) {
+            this.drawCenteredString(this.fontRenderer, lines.get(i), this.width / 2, 10 + i * 10, 0xFF7F7F7F);
+        }
+        super.drawScreen(mouseX, mouseY, renderPartialTick);
+    }
+}
