--- /dev/null
+++ b/io/github/bluestaggo/mcdiverge/GuiRefabricator.java
@@ -1,0 +1,43 @@
+package io.github.bluestaggo.mcdiverge;
+
+import net.minecraft.src.*;
+import org.lwjgl.opengl.GL11;
+
+public class GuiRefabricator extends GuiContainer {
+    private TileEntityRefabricator refabricator;
+
+    public GuiRefabricator(InventoryPlayer inventoryPlayer1, TileEntityRefabricator tileEntityRefabricator) {
+        this.refabricator = tileEntityRefabricator;
+        this.inventorySlots.add(new SlotInventory(this, tileEntityRefabricator, 0, 80, 35));
+        this.inventorySlots.add(new SlotInventory(this, tileEntityRefabricator, 1, 107, 58));
+
+        int i3;
+        for(i3 = 0; i3 < 3; ++i3) {
+            for(int i4 = 0; i4 < 9; ++i4) {
+                this.inventorySlots.add(new SlotInventory(this, inventoryPlayer1, i4 + (i3 + 1) * 9, 8 + i4 * 18, 84 + i3 * 18));
+            }
+        }
+
+        for(i3 = 0; i3 < 9; ++i3) {
+            this.inventorySlots.add(new SlotInventory(this, inventoryPlayer1, i3, 8 + i3 * 18, 142));
+        }
+
+    }
+
+    protected void drawGuiContainerForegroundLayer() {
+        this.fontRenderer.drawString("Refabricator", 60, 6, 4210752);
+        this.fontRenderer.drawString("Inventory", 8, this.ySize - 96 + 2, 4210752);
+    }
+
+    protected void drawGuiContainerBackgroundLayer(float renderPartialTick) {
+        int bg = this.mc.renderEngine.getTexture("/gui/refabricator.png");
+        GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F);
+        this.mc.renderEngine.bindTexture(bg);
+        int posX = (this.width - this.xSize) / 2;
+        int posY = (this.height - this.ySize) / 2;
+        this.drawTexturedModalRect(posX, posY, 0, 0, this.xSize, this.ySize);
+
+        int i5 = this.refabricator.getProgressScaled(48);
+        this.drawTexturedModalRect(posX + 55, posY + 26 + (48 - i5), 176, 48 - i5, 16, i5);
+    }
+}
