--- /dev/null
+++ b/io/github/bluestaggo/mcdiverge/ItemCharm.java
@@ -1,0 +1,31 @@
+package io.github.bluestaggo.mcdiverge;
+
+import net.minecraft.src.EntityPlayer;
+import net.minecraft.src.ItemEquipable;
+import net.minecraft.src.ItemStack;
+
+public class ItemCharm extends ItemEquipable {
+    public final String textureName;
+
+    public ItemCharm(int id) {
+        super(id, 4);
+        this.textureName = null;
+    }
+
+    public ItemCharm(int id, String renderName) {
+        super(id, 4);
+        this.textureName = renderName;
+    }
+
+    public ItemCharm(int id, int durability) {
+        super(id, 4, durability);
+        this.textureName = null;
+    }
+
+    public ItemCharm(int id, int durability, String renderName) {
+        super(id, 4, durability);
+        this.textureName = renderName;
+    }
+
+    public void updateCharm(EntityPlayer player, ItemStack stack) {}
+}
