--- /dev/null
+++ b/io/github/bluestaggo/mcdiverge/ItemRepairKit.java
@@ -1,0 +1,31 @@
+package io.github.bluestaggo.mcdiverge;
+
+import net.minecraft.src.*;
+
+public class ItemRepairKit extends ItemCharm {
+    public ItemRepairKit(int id) {
+        super(id, 1024);
+    }
+
+    public void updateCharm(EntityPlayer player, ItemStack stack) {
+        if (player.isSneaking()) {
+            ItemStack heldItem = player.inventory.getCurrentItem();
+            if (rand.nextInt(2) != 0
+                || heldItem == null
+                || heldItem.itemDmg == 0
+                || !Item.itemsList[heldItem.itemID].allowRepair()) {
+                return;
+            }
+            --heldItem.itemDmg;
+            stack.damageItem(1);
+            player.playSound("random.wood click", 1.0F, 1.0F);
+        }
+    }
+
+    public Object[] getDynamicTooltip() {
+        return new Object[] {
+                this.maxDamage + " max durability",
+                ItemTooltip.INFO_COL
+        };
+    }
+}
