--- /dev/null
+++ b/io/github/bluestaggo/mcdiverge/WorldGenSpeleothems.java
@@ -1,0 +1,50 @@
+package io.github.bluestaggo.mcdiverge;
+
+import net.minecraft.src.World;
+import net.minecraft.src.WorldGenerator;
+
+import java.util.Random;
+
+public class WorldGenSpeleothems extends WorldGenerator {
+    private final int blockId;
+
+    public WorldGenSpeleothems(int blockId) {
+        this.blockId = blockId;
+    }
+
+    public boolean generate(World world1, Random random2, int i3, int i4, int i5) {
+        boolean stalactite = random2.nextInt(2) == 0;
+        if (stalactite) {
+            i4 = findStalactiteAnchor(world1, i3, i4, i5);
+        } else {
+            i4 = findStalagmiteAnchor(world1, i3, i4, i5);
+        }
+        if (i4 == -1 || i4 > 64) {
+            return false;
+        }
+        int height = random2.nextInt(2) + 1;
+        for (int i = 0; i < height; i++) {
+            i4 += stalactite ? -1 : 1;
+            world1.setBlockWithNotify(i3, i4, i5, blockId);
+        }
+        return true;
+    }
+
+    private int findStalactiteAnchor(World world1, int x, int y, int z) {
+        for (++y; y < 128; y++) {
+            if (world1.isBlockNormalCube(x, y, z) && world1.getBlockId(x, y - 1, z) == 0) {
+                return y;
+            }
+        }
+        return -1;
+    }
+
+    private int findStalagmiteAnchor(World world1, int x, int y, int z) {
+        for (--y; y > 0; y--) {
+            if (world1.isBlockNormalCube(x, y, z) && world1.getBlockId(x, y + 1, z) == 0) {
+                return y;
+            }
+        }
+        return -1;
+    }
+}
