--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -12,28 +12,7 @@
 import java.util.logging.Level;
 import java.util.logging.Logger;
 
-import net.minecraft.src.AxisAlignedBB;
-import net.minecraft.src.ConsoleLogManager;
-import net.minecraft.src.EntityPlayerMP;
-import net.minecraft.src.EntityTracker;
-import net.minecraft.src.ICommandListener;
-import net.minecraft.src.IProgressUpdate;
-import net.minecraft.src.IUpdatePlayerListBox;
-import net.minecraft.src.Item;
-import net.minecraft.src.ItemStack;
-import net.minecraft.src.NetworkListenThread;
-import net.minecraft.src.Packet3Chat;
-import net.minecraft.src.Packet4UpdateTime;
-import net.minecraft.src.PropertyManager;
-import net.minecraft.src.ServerCommand;
-import net.minecraft.src.ServerConfigurationManager;
-import net.minecraft.src.ServerGUI;
-import net.minecraft.src.ThreadCommandReader;
-import net.minecraft.src.ThreadServerApplication;
-import net.minecraft.src.ThreadSleepForeverServer;
-import net.minecraft.src.Vec3D;
-import net.minecraft.src.WorldManager;
-import net.minecraft.src.WorldServer;
+import net.minecraft.src.*;
 
 public class MinecraftServer implements ICommandListener, Runnable {
     public static Logger logger = Logger.getLogger("Minecraft");
@@ -106,7 +85,7 @@
 
     private void initWorld(String worldName) {
         logger.info("Preparing start region");
-        this.worldMngr = new WorldServer(new File("."), worldName, this.propertyManagerObj.getBooleanProperty("monsters", false));
+        this.worldMngr = new WorldServer(new File("."), worldName, this.propertyManagerObj.getBooleanProperty("monsters", false), this.propertyManagerObj.getCreateWorldInfo());
         this.worldMngr.addWorldAccess(new WorldManager(this));
         this.worldMngr.difficultySetting = 1;
         this.configManager.setPlayerManager(this.worldMngr);
@@ -265,7 +244,6 @@
         } catch (Exception exception5) {
             logger.log(Level.WARNING, "Unexpected exception while parsing console command", exception5);
         }
-
     }
 
     public void addCommand(String command, ICommandListener commandListener) {
@@ -314,6 +292,16 @@
                         string11 = string2.substring(string2.indexOf(" ")).trim();
                         this.configManager.pardonIP(string11);
                         this.print(string4, "Pardoning ip " + string11);
+                    } else if(string2.toLowerCase().startsWith("creative ")) {
+                        string11 = string2.substring(string2.indexOf(" ")).trim();
+                        this.configManager.setPlayerCreative(string11, true);
+                        this.print(string4, "Setting " + string11 + "'s gamemode into creative");
+                        this.configManager.sendChatMessageToPlayer(string11, "\u00a7eYou are now in creative mode!");
+                    } else if(string2.toLowerCase().startsWith("survival ")) {
+                        string11 = string2.substring(string2.indexOf(" ")).trim();
+                        this.configManager.setPlayerCreative(string11, false);
+                        this.print(string4, "Setting " + string11 + "'s gamemode into survival");
+                        this.configManager.sendChatMessageToPlayer(string11, "\u00a7eYou are now in survival mode!");
                     } else {
                         EntityPlayerMP entityPlayerMP12;
                         if(string2.toLowerCase().startsWith("ban ")) {
@@ -444,6 +432,8 @@
                 iCommandListener3.addHelpCommandMessage("   save-on                   re-enables terrain saving");
                 iCommandListener3.addHelpCommandMessage("   list                      lists all currently connected players");
                 iCommandListener3.addHelpCommandMessage("   say <message>             broadcasts a message to all players");
+                iCommandListener3.addHelpCommandMessage("   creative <player>         puts a player into creative mode");
+                iCommandListener3.addHelpCommandMessage("   survival <player>         puts a player into survival mode");
             }
         }
 
